rust/sbd-client/src/send_buf.rs

```diff
@@ -0,0 +1,159 @@
+use super::*;
+
+use std::collections::VecDeque;
+
+pub struct SendBuf {
+    pub ws: raw_client::WsRawSend,
+    pub buf: VecDeque<Vec<u8>>,
+    pub out_buffer_size: usize,
+    pub origin: tokio::time::Instant,
+    pub limit_rate: u64,
+    pub idle_keepalive_nanos: u64,
+    pub next_send_at: u64,
+    pub last_send: u64,
+}
+
+impl SendBuf {
+    /// construct a new send buf
+    pub fn new(
+        ws: raw_client::WsRawSend,
+        out_buffer_size: usize,
+        limit_rate: u64,
+        idle_keepalive: std::time::Duration,
+        pre_sent_bytes: usize,
+    ) -> Self {
+        let mut this = Self {
+            ws,
+            buf: VecDeque::default(),
+            out_buffer_size,
+            origin: tokio::time::Instant::now(),
+            limit_rate,
+            idle_keepalive_nanos: idle_keepalive.as_nanos() as u64,
+            next_send_at: 0,
+            last_send: 0,
+        };
+
+        let now = this.origin.elapsed().as_nanos() as u64;
+
+        this.next_send_at = std::cmp::max(now, this.next_send_at)
+            + (pre_sent_bytes as u64 * this.limit_rate);
+
+        this
+    }
+
+    /// Close the connection.
+    pub async fn close(&mut self) {
+        self.ws.close().await;
+    }
+
+    /// We received a new rate limit from the server, update our records.
+    pub fn new_rate_limit(&mut self, limit: u64) {
+        if limit < self.limit_rate {
+            // rate limit updates are sent on a best effort,
+            // and there are network timing conditions to worry about.
+            // assume we accidentally sent a message while the new limit
+            // was in effect, and accout for that in a brute-force manner.
+
+            let now = self.origin.elapsed().as_nanos() as u64;
+
+            self.next_send_at = std::cmp::max(now, self.next_send_at)
+                + (MAX_MSG_SIZE as u64 * self.limit_rate);
```

Comment:
would this be enough?

```suggestion
                + (self.len() * self.limit_rate);
```

Reply:
mmm, that would make it very variable... anywhere between zero and the full buffer size (8 * MAX_MSG_SIZE currently). And the buffer size has nothing to do with how many bytes we've sent since the server decided to divide the bandwidth up differently.

I'm not sure the current solution is the best, but I think self.len() might be a further departure.